### PR TITLE
PHP 8 support

### LINKED ISSUE #23: [PHP8] ArrayObject::asort
PHP8 adds the $flag param to [ArrayObject::asort](https://www.php.net/manual/en/arrayobject.asort.php)

`public ArrayObject::asort ( int $flags = SORT_REGULAR ) : void`

This causes a Fatal Error when trying to use this method.

`PHP Fatal error:  Declaration of Managur\Collection\Collection::asort(?callable $callable = null): Managur\Collection\Collection must be compatible with ArrayObject::asort(int $flags = SORT_REGULAR) in /app/vendor/managur/collection/src/Collection.php on line 393`

Need to implement the $flags param and handle the additional behaviours of the flags.

- PHP version - 8.0.1
- Managur/collection version - 1.7.2

### PR DESCRIPTION
**Pull request by Antnee**
Closes #23

### CODE REVIEW

**Comment by Antnee** [comment]
Interested in your thoughts on this now, @lewischadwick

File: src/Collection.php [hunk 1]
```diff
@@ -362,52 +382,77 @@ public function merge(Collection $add): self
...
-            }
+        if ($this->keyType) {
+            asort($data, $flags);
         } else {
-            if ($this->keyType) {
-                asort($data);
-            } else {
-                sort($data);
-            }
+            sort($data);
```
Comment (lewischadwick) [inline src/Collection.php]: Should this also take `$flags`?

File: src/Collection.php [hunk 2]
```diff
@@ -454,32 +500,34 @@ private function getNewInstance($data): self
...
+     * @param ?string $keyType
+     * @param ?string $valueType
      * @return self
      */
-    private static function getTypedCollection($data, string $keyType = null, string $valueType = null): Collection
-    {
-        $ret = new class ($data, $keyType, $valueType) extends Collection {
+    private static function getTypedCollection(
+        mixed $data,
+        string $keyType = null,
```
Comment (lewischadwick) [inline src/Collection.php]: Should this be `?string`?

File: src/Collection.php [hunk 3]
```diff
@@ -454,32 +500,34 @@ private function getNewInstance($data): self
...
+        string $valueType = null,
```
Comment (lewischadwick) [inline src/Collection.php]: Should this be `?string`?

File: src/Collection.php [hunk 4]
```diff
@@ -209,13 +222,18 @@ public function map(callable $callable): self
...
+     *                      present.
+     *                      If length is given and is negative then the sequence will stop that many elements from the
+     *                      end of the array.
+     *                      If it is omitted, then the sequence will have everything from offset up until the end of the
+     *                      array.
      * @return static New collection of the same type
      */
-    public function slice(int $offset, int $length = null): self
+    // phpcs:ignore Squiz.WhiteSpace.ScopeKeywordSpacing.Incorrect -- Broken until 3.6.0
+    public function slice(int $offset, int $length = null): static
```
Comment (lewischadwick) [inline src/Collection.php]: should this be `?int length = null`?

File: src/Collection.php [hunk 5]
```diff
@@ -253,7 +272,8 @@ public function reduce(callable $callable, $carry = null)
      * @param int|null $flag Collection::FILTER_USE_KEY or Collection::FILTER_USE_BOTH
      * @return static
      */
-    public function filter($callable = null, int $flag = null): self
+    // phpcs:ignore Squiz.WhiteSpace.ScopeKeywordSpacing.Incorrect -- Broken until 3.6.0
+    public function filter($callable = null, int $flag = null): static
```
Comment (lewischadwick) [inline src/Collection.php]: should this be `?int $flag = null`?

File: src/Collection.php [hunk 6]
```diff
@@ -292,9 +311,9 @@ public function first(callable $callable = null, $default = null)
      * @param mixed|null If no result is found, return this instead
      * @return mixed
      */
-    public function last(callable $callable = null, $default = null)
+    public function last(callable $callable = null, $default = null): mixed
```
Comment (lewischadwick) [inline src/Collection.php]: should this be `?callable $callable = null, mixed $default = null): mixed`?

File: src/Collection.php [hunk 7]
```diff
@@ -236,11 +254,12 @@ public function each(callable $callable)
...
      * Reduce Collection by Callable
      *
-     * @param callable $callable Requires two arguments; the first to carry from the previous iteration, and the second as the item
+     * @param callable $callable Requires two arguments; the first to carry from the previous iteration, and the second
+     *                           as the item
      * @param mixed $carry Initial value, or returned if array is empty
      * @return mixed Type depends on return value of $callable
      */
-    public function reduce(callable $callable, $carry = null)
+    public function reduce(callable $callable, $carry = null): mixed
```
Comment (lewischadwick) [inline src/Collection.php]: Should this be `mixed $carry = null`?

File: src/Collection.php (hunk 5 above)
Comment (Antnee) [reply src/Collection.php]: Actually `?callable $callable = null, ?int $flag = null` by the looks of it 😄

**Review by lewischadwick — APPROVED** [review]

**Pull request merged**